### ISSUE TITLE
OutOfMemoryException Issues

### ISSUE THREAD

**Issue by gmoreira** [ekino/EkinoNewRelicBundle#86]
I am occasionally seeing this error from this bundle's ResponseListener:

`Uncaught PHP Exception Symfony\Component\Debug\Exception\OutOfMemoryException: "Error: Allowed memory size of 134217728 bytes exhausted (tried to allocate 17538363 bytes)" at /somepath/.../vendor/ekino/newrelic-bundle/Ekino/Bundle/NewRelicBundle/Listener/ResponseListener.php line 86`

It would seem that with lengthy HTML pages, preg_replace is not very memory efficient and causes this exception.

I know that I can simply increase the PHP memory limit per request to address this, but this shouldn't be necessary.

Are there any ideas on how we can address this with a more efficient solution than parsing the HTML response each time? Using a Twig extension perhaps as an alternate approach?

**Comment by gmoreira** [comment]
Is something like this an option perhaps?
http://php.net/manual/en/function.preg-replace.php#84285

**Comment by rande** [comment]
Cannot you just raise the memory limit ?

**Comment by gmoreira** [comment]
@rande - sure, but increasing memory is a short term fix that doesn't actually address the core problem here - the inefficiency in the approach.

**Comment by Rajat001** [comment]
How u had solved this problem @gmoreira  , i also had the same prob ...
Is there need to run an external command on command prompt or doing external action outside the cmd ..

**Comment by gmoreira** [comment]
@Rajat001 no I haven't. They only work around I have found is to reduce the HTML footprint of the pages in question affected by this issue.

**Comment by jderusse** [comment]
The memory in consumed by the listener when it tries to inject instrumentation code in the response. The current implementation copy the HTML response in a variable to inject the code in the header and the footer that's why during few seconds the response is duplicate in 2 variables. The only way to fix it is:
- increase memory_limit
- reduce response size
- disable `_instrument` option

**Comment by Nyholm** [comment]
Maybe substr will be more memory efficient. See example from the symfony toolbar. 

https://github.com/symfony/symfony/blob/master/src/Symfony/Bundle/WebProfilerBundle/EventListener/WebDebugToolbarListener.php

**Comment by jderusse** [comment]
This is not related to the preg_replace or substr, the issue is: by modifying the variable ($content) php stop pointing to the reference of the same memory address, and copy the variable to a new location. The response is stored in 2 place (in the Response object AND in the $content variable). This mlemory is freed when the new content is set back in the response.

Here is a simple test

```
$str = str_repeat(' ', 1000);
$response->setContent($str);
unset($str);

$mem['start'] = memory_get_usage();
$content = $response->getContent();
$mem['get'] = memory_get_usage();
$content .= '-';
$mem['modified'] = memory_get_usage();
$response->setContent($content);
$mem['set'] = memory_get_usage();

echo 'get: '.($mem['get']-$mem['start']) . PHP_EOL; // 376
echo 'modified: '.($mem['modified']-$mem['get']) . PHP_EOL; // 1280
echo 'set: '.($mem['set']-$mem['modified']) . PHP_EOL; // -1280
```

I think that the solution is to temporary inject a blank content to the response, which would release the old variable and free memory
```
$str = str_repeat(' ', 1000);
$response->setContent($str);
unset($str);

$mem['start'] = memory_get_usage();
$content = $response->getContent();
$mem['get'] = memory_get_usage();
$response->setContent(''); // <===== HERE
$content[0] = '-';
$mem['modified'] = memory_get_usage();
$response->setContent($content);
$mem['set'] = memory_get_usage();

echo 'get: '.($mem['get']-$mem['start']) . PHP_EOL; // 376
echo 'modified: '.($mem['modified']-$mem['get']) . PHP_EOL; // 0
echo 'set: '.($mem['set']-$mem['modified']) . PHP_EOL; // 0
```

**Comment by Nyholm** [comment]
Im closing this as it was fixed in #191
